s3s3/api.py

```diff
@@ -0,0 +1,114 @@
+
+
+def create_connection(connection_args):
+    """
+    Create a boto s3 connection using the connection_args dictionary.
+    ``connection_args`` A dictionary of connection arguments.
+    """
+    connection_args = connection_args.copy()
+    # bucket name is an invalid connection argument.
+    connection_args.pop('bucket_name')
+    connection_args.pop('verify_md5')
+    return S3Connection(**connection_args)
+
+
+def upload(source_key, dest_keys, verify_md5=False):
+    """
+    Download from the source s3 key to a named temporary file
+    and upload to the destination s3 key.
+    ``source_key`` The source boto s3 key.
+    ``dest_keys`` A list of the destination boto s3 keys.
+    ``verify_md5`` Verify md5 values when uploading. Source is always
+    considered authoritative.
+    """
+    # Use the same name if no destination key is passed.
+    if not dest_keys or not source_key:
+        raise Exception(
+            'The source_key and dest_keys parameters are required.')
+    with tempfile.NamedTemporaryFile() as data:
+        source_key.get_contents_to_file(data)
+        data.file.flush()
+        for dest_key in dest_keys:
+            dest_key.set_contents_from_filename(data.name)
+            _upload_verify_md5(verify_md5, source_key, dest_key)
+            try:
+                r.set(u'backup=>' + dest_key.key, True)
+            except redis.ConnectionError:
+                logging.warn('Unable to connect to redis')
+
+
+def duplicate_bucket(source_bucket, dest_bucket, verify_md5=False):
+    """
+    Duplicate the source s3 bucket in the destination bucket.
+    ``source_bucket`` The source boto s3 bucket.
+    ``dest_bucket`` The destination boto s3 bucket.
+    ``verify_md5`` Verify md5 values when duplicating. Source always
+                   is considered authoritative. If md5 is missing this
+                   will fetch the file to compute the md5 and make the
+                   comparison.
+    """
+    for source_key in source_bucket.get_all_keys():
+        dest_key = Key(dest_bucket)
+        dest_key.key = source_key.key
+        if not dest_key.exists():
+            upload(source_key, [dest_key])
+        if verify_md5:
+            _update_md5([source_key, dest_key])
+            if source_key.md5 != dest_key.md5:
+                upload(source_key, [dest_key])
+                logging.info('Uploaded {0} to {1} in bucket {2}.'.format(
+                    dest_key.key,
+                    dest_key.bucket.connection.host,
+                    dest_key.bucket.name))
+
+
+def _update_md5(keys):
+    """
+    If the md5 does not exist then retrieve the file so the md5 can be
+    computed. This is required because some s3 clones do not provide
+    md5 values.
+
+    ``keys`` boto.s3.key.Key objects which may not have md5 hashes.
+    """
+    for key in keys:
+        if not key.md5:
+            with tempfile.NamedTemporaryFile() as data:
+                key.get_contents_to_file(data)
```

Comment:
The code for [boto.s3.key.Key.md5 (boto==2.38.0)](https://github.com/boto/boto/blob/2.38.0/boto/s3/key.py#L167-L177).

Specifically the [code that updates md5 (and other digesters) is in `_get_file_internal`](https://github.com/boto/boto/blob/2.38.0/boto/s3/key.py#L1554-L1555).

The problem is as seen from the ipython:

```
In [23]: key._get_remote_metadata()
Out[23]: {'content-type': 'application/octet-stream'}
```
Ceph doesn't serve up the content-md5 in it's response.

```
curl -i https://s3.lsst.codes/lsst-git-lfs/data/f6b0597622934ba80fabc1d1c6afb83004215e9fc00432350444c18082030d2f -o f6b0597622934ba80fabc1d1c6afb83004215e9fc00432350444c18082030d2f
http://docs.ceph.com/docs/master/radosgw/s3/objectops/
```

which has these response headers.

```
HTTP/1.1 200 OK
Date: Thu, 29 Oct 2015 04:02:46 GMT
Server: Apache/2.4.7 (Ubuntu)
Accept-Ranges: bytes
Strict-Transport-Security: max-age=63072000; includeSubDomains
X-Frame-Options: SAMEORIGIN
Content-Length: 98150412
Last-Modified: Tue, 20 Oct 2015 17:09:48 GMT
ETag: "8cbd36bc39ab5e0c4ad0e8469534c5e5"
Public-Key-Pins: pin-sha256="<Subject Public Key Information (SPKI)>"; max-age=2592000; includeSubDomains
Content-Type: application/octet-stream
```
More docs about what Ceph implements.

http://docs.ceph.com/docs/master/radosgw/s3/objectops/
http://docs.ceph.com/docs/master/dev/radosgw/s3_compliance/